x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlProcessorAutoscalingDecider.java

```diff
@@ -0,0 +1,160 @@
+            .setReason(
+                String.format(
+                    Locale.ROOT,
+                    "Passing currently perceived capacity as down scale delay has not been satisfied; configured delay [%s] "
+                        + "last detected scale down event [%s]. Will request scale down in approximately [%s]",
+                    downScaleDelay.getStringRep(),
+                    XContentElasticsearchExtension.DEFAULT_FORMATTER.format(Instant.ofEpochMilli(scaleTimer.downScaleDetectedMillis())),
+                    TimeValue.timeValueMillis(msLeftToScale).getStringRep()
+                )
+            )
+            .build();
+    }
+
+    private boolean hasUnsatisfiedDeployments(TrainedModelAssignmentMetadata trainedModelAssignmentMetadata, List<DiscoveryNode> mlNodes) {
+        final Set<String> mlNodeIds = mlNodes.stream().map(DiscoveryNode::getId).collect(Collectors.toSet());
+        return trainedModelAssignmentMetadata.modelAssignments()
+            .values()
+            .stream()
+            .anyMatch(deployment -> deployment.isSatisfied(mlNodeIds) == false);
+    }
+
+    private MlProcessorAutoscalingCapacity.Builder computeRequiredCapacity(TrainedModelAssignmentMetadata trainedModelAssignmentMetadata) {
+        int maxThreadsPerAllocation = 0;
+        int processorCount = 0;
+        for (TrainedModelAssignment assignment : trainedModelAssignmentMetadata.modelAssignments().values()) {
+            int threadsPerAllocation = assignment.getTaskParams().getThreadsPerAllocation();
+            maxThreadsPerAllocation = Math.max(maxThreadsPerAllocation, threadsPerAllocation);
+            processorCount += assignment.getTaskParams().getNumberOfAllocations() * threadsPerAllocation;
+        }
+
+        final int numMlAvailabilityZones = nodeAvailabilityZoneMapper.getNumMlAvailabilityZones().orElse(1);
+        if (numMlAvailabilityZones > 1) {
+            // We assume cloud provides what we ask for tier processors for each availability zone.
+            // Thus we need to devide the total processor count required by the number of ML availability zones.
+            processorCount = (processorCount - 1) / numMlAvailabilityZones + 1;
+        }
+        processorCount = Math.max(processorCount, maxThreadsPerAllocation);
+
+        return MlProcessorAutoscalingCapacity.builder(maxThreadsPerAllocation, processorCount);
+    }
+
+    MlProcessorAutoscalingCapacity computeCurrentCapacity(List<DiscoveryNode> mlNodes) {
+        int maxNodeProcessors = 0;
+        int tierProcessors = 0;
+        for (DiscoveryNode node : mlNodes) {
+            int nodeProcessors = getProcessors(node);
+            maxNodeProcessors = Math.max(maxNodeProcessors, nodeProcessors);
+            tierProcessors += nodeProcessors;
+        }
+        return MlProcessorAutoscalingCapacity.builder(maxNodeProcessors, tierProcessors).build();
+    }
```

Comment:
I think there could be a nasty interaction with #89281 here.  #89281 allows `node.processors` to be a `double`, but then rounds the number up when passing it on as an `int` to consumers that expected an `int` before.

But this could lead to spurious scale ups in Cloud.  Suppose we have 3 AZs in the ML tier, and 3 ML nodes, one per AZ, and a CPU quota of 2.5 per node.  So the whole ML tier has a total CPU quota of 7.5.  But with the rounding of 2.5 to 3 for each node we will ask for 9 CPUs for the tier when we mean "no change required".  This will lead to an unnecessary scale up.  Even at the larger tier size the same thing could happen again if `node.processors` is still fractional.  We'd keep asking for scale ups until `node.processors` was set to an integer.

I think that in places where `node.processors` is used in autoscaling code we'll need to get and use the floating point value rather than the rounded up value, so we'll need a new method `EsExecutors.unroundedAllocatedProcessors()` (or some similar name) and use that within autoscaling code.

Do you have any thoughts on this @fcofdez or @henningandersen?

Reply:
#89662 should address this I think. Hopefully we can get that one in soon and unblock this.

However, this does make your code dependent on receiving the right input from the orchestration system.